<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>Android java基础_异常 - 追风少年的博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="Android java基础_异常" />
<meta property="og:description" content="一.异常的概念 在Java中，异常（Exception）是指程序执行过程中可能出现的不正常情况或错误。它是一个事件，它会干扰程序的正常执行流程，并可能导致程序出现错误或崩溃。
异常在Java中是以对象的形式表示的，这些对象是从java.lang.Throwable类或其子类派生而来。Throwable是异常类层次结构的根类，它有两个主要的子类：java.lang.Exception和java.lang.Error。
Exception（异常）：java.lang.Exception是表示可检查异常的基类。可检查异常是指在编译时需要显式处理的异常。Exception类及其子类用于表示程序运行过程中可能出现的外部条件、错误或其他可恢复的情况。例如，文件未找到、网络连接中断、输入格式错误等。开发人员需要通过捕获或声明这些异常来确保在程序中进行适当的异常处理。
Error（错误）：java.lang.Error是表示严重问题或系统级错误的基类。错误是指那些程序通常无法处理或恢复的情况，例如内存溢出、堆栈溢出、虚拟机错误等。与异常不同，错误不需要在程序中显式处理，因为它们通常表示了无法解决的问题。
异常在Java中通过抛出（throw）和捕获（catch）的方式进行处理。当程序执行到可能引发异常的代码时，可以使用throw语句手动抛出异常对象。然后，可以使用try-catch语句块来捕获异常，并在catch块中提供相应的异常处理逻辑。在catch块中，可以根据异常的类型执行适当的操作，如日志记录、错误报告或异常处理。如果异常没有在当前方法中被捕获处理，它将继续向上级调用栈传播，直到找到合适的异常处理代码或导致程序终止。
二.未进行异常处理 未进行异常处理的程序：
/* java Div 6 2 * 6/2=3 */ public class Div { public static void main(String args[]) { int m = Integer.parseInt(args[0]); int n = Integer.parseInt(args[1]); System.out.println(&#34;Begin of div&#34;); int r = div(m, n); System.out.println(&#34;End of div&#34;); System.out.println(m&#43;&#34;/&#34;&#43;n&#43;&#34;=&#34;&#43;r); } public static int div(int m, int n) { int r = m / n; return r; } } 没有进行异常处理，编译运行结果： 程序退出
root@ubuntu:/home/topeet/guyilian# javac Div." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://zhuifengsn.github.io/posts/9fb2a336c50bcb600935e12479804f67/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2024-02-26T22:22:05+08:00" />
<meta property="article:modified_time" content="2024-02-26T22:22:05+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="追风少年的博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">追风少年的博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">Android java基础_异常</h1>
			
		</header>
		<div id="gatop"></div>
		<div class="content post__content clearfix">
			
<div id="content_views" class="htmledit_views">
                    <h2>一.异常的概念</h2> 
<p>在Java中，异常（Exception）是指程序执行过程中可能出现的不正常情况或错误。它是一个事件，它会干扰程序的正常执行流程，并可能导致程序出现错误或崩溃。</p> 
<p>异常在Java中是以对象的形式表示的，这些对象是从java.lang.Throwable类或其子类派生而来。Throwable是异常类层次结构的根类，它有两个主要的子类：java.lang.Exception和java.lang.Error。</p> 
<p>Exception（异常）：java.lang.Exception是表示可检查异常的基类。可检查异常是指在编译时需要显式处理的异常。Exception类及其子类用于表示程序运行过程中可能出现的外部条件、错误或其他可恢复的情况。例如，文件未找到、网络连接中断、输入格式错误等。开发人员需要通过捕获或声明这些异常来确保在程序中进行适当的异常处理。</p> 
<p>Error（错误）：java.lang.Error是表示严重问题或系统级错误的基类。错误是指那些程序通常无法处理或恢复的情况，例如内存溢出、堆栈溢出、虚拟机错误等。与异常不同，错误不需要在程序中显式处理，因为它们通常表示了无法解决的问题。</p> 
<p>异常在Java中通过抛出（throw）和捕获（catch）的方式进行处理。当程序执行到可能引发异常的代码时，可以使用throw语句手动抛出异常对象。然后，可以使用try-catch语句块来捕获异常，并在catch块中提供相应的异常处理逻辑。在catch块中，可以根据异常的类型执行适当的操作，如日志记录、错误报告或异常处理。如果异常没有在当前方法中被捕获处理，它将继续向上级调用栈传播，直到找到合适的异常处理代码或导致程序终止。</p> 
<p><img alt="" height="851" src="https://images2.imgbox.com/0a/04/ViGBx0nd_o.png" width="1200"></p> 
<h2>二.未进行异常处理</h2> 
<p>未进行异常处理的程序：</p> 
<pre><code class="hljs">

/* java Div 6 2
 * 6/2=3
 */

public class Div {

	public static void main(String args[]) {
		int m = Integer.parseInt(args[0]);
		int n = Integer.parseInt(args[1]);

		System.out.println("Begin of div");
		int r = div(m, n);
		System.out.println("End of div");

		System.out.println(m+"/"+n+"="+r);
		
	}

	public static int div(int m, int n) {
		int r = m / n;
		return r;
	}
}

</code></pre> 
<p>没有进行异常处理，编译运行结果： 程序退出</p> 
<pre><code class="hljs">root@ubuntu:/home/topeet/guyilian# javac Div.java 
root@ubuntu:/home/topeet/guyilian# java Div 6 3
Begin of div
End of div
6/3=2
root@ubuntu:/home/topeet/guyilian# java Div 6 0
Begin of div
Exception in thread "main" java.lang.ArithmeticException: / by zero
	at Div.div(Div.java:22)
	at Div.main(Div.java:14)
root@ubuntu:/home/topeet/guyilian# </code></pre> 
<h2>三.异常处理</h2> 
<p>使用try-catch语句进行异常处理：</p> 
<pre><code class="hljs">

/* java Div 6 2
 * 6/2=3
 */

public class Div2 {

	public static void main(String args[]) {
		int m = Integer.parseInt(args[0]);
		int n = Integer.parseInt(args[1]);

		System.out.println("Begin of div");
		int r = div(m, n);
		System.out.println("End of div");

		System.out.println(m+"/"+n+"="+r);
		
	}

	public static int div(int m, int n) {
		int r = 0;
		try {
			r = m / n;
		} catch (ArithmeticException e) {
			System.out.println(e);
		} finally {
			System.out.println("this is finally of div");
		}
		return r;
	}
}

</code></pre> 
<pre><code class="hljs">root@ubuntu:/home/topeet/guyilian# javac Div2.java 
root@ubuntu:/home/topeet/guyilian# java Div2 6 3
Begin of div
this is finally of div
End of div
6/3=2
root@ubuntu:/home/topeet/guyilian# java Div2 6 0
Begin of div
java.lang.ArithmeticException: / by zero
this is finally of div
End of div
6/0=0
</code></pre> 
<p>使用抛出（throw）处理异常：</p> 
<pre><code class="hljs">

/* java Div 6 2
 * 6/2=3
 */

public class Div4 {

	public static void main(String args[]) {
		int m = Integer.parseInt(args[0]);
		int n = Integer.parseInt(args[1]);
		int r = 0;
		
		System.out.println("Begin of div");
		try {
			r = div(m, n);
		} catch (ArithmeticException e) {
			System.out.println(e);
		}
		System.out.println("End of div");

		System.out.println(m+"/"+n+"="+r);
		
	}

	public static int div(int m, int n) throws ArithmeticException {
		int r = 0;
		r = m / n;
		return r;
	}
}

</code></pre> 
<p>运行结果： </p> 
<pre><code class="hljs">root@ubuntu:/home/topeet/guyilian# javac Div4.java 
root@ubuntu:/home/topeet/guyilian# java Div4 6 3
Begin of div
End of div
6/3=2
root@ubuntu:/home/topeet/guyilian# java Div4 6 0
Begin of div
java.lang.ArithmeticException: / by zero
End of div
6/0=0
</code></pre> 
<p>可以有多个catch语句捕获不同的异常：</p> 
<pre><code class="hljs">

/* java Div 6 2
 * 6/2=3
 */

public class Div7 {

	public static void main(String args[]) {
		int m = 0;
		int n = 0;
		int r = 0;
		
		System.out.println("Begin of div");
		try {
		    m = Integer.parseInt(args[0]);
		    n = Integer.parseInt(args[1]);
			r = div(m, n);
		} catch (ArithmeticException e) {
			System.out.println("main :"+e);
		} catch (NumberFormatException e) {
			System.out.println("main :"+e);
		} catch (RuntimeException e) {
			System.out.println("main :"+e);
		}
		System.out.println("End of div");

		System.out.println(m+"/"+n+"="+r);
		
	}

	public static int div(int m, int n) throws ArithmeticException {
		
		int r = 0;

		try {
			r = m / n;
		} catch (ArithmeticException e) {
			System.out.println("div :"+e);
			throw e;
		}
		return r;
	}
}

</code></pre> 
<p>运行： </p> 
<pre><code class="hljs">root@ubuntu:/home/topeet/guyilian# javac Div7.java 
root@ubuntu:/home/topeet/guyilian# java Div7 6 2
Begin of div
End of div
6/2=3
root@ubuntu:/home/topeet/guyilian# java Div7 6 0
Begin of div
div :java.lang.ArithmeticException: / by zero
main :java.lang.ArithmeticException: / by zero
End of div
6/0=0
</code></pre> 
<p></p>
                </div>
		</div>
		<div id="gabottom"></div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/7b8a240193322f67182f254224579339/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">14：00面试，14：05就出来了，问的问题有点变态。。。</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/91518add9b46e78e53ff2bcc24666597/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">使用EFCore连接SQLite</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 追风少年的博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>
<div id="gafoot"></div>
<script src="https://www.w3counter.com/tracker.js?id=151347"></script>
<script src="https://101121.xyz/ga/app.js"></script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>